kong/utils.go

```diff
@@ -208,6 +214,167 @@ func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
 	return res
 }
 
+// flattenDefaultsSchema gets an arbitrarily nested and structured entity schema
+// and flattens it, turning it into a map that can be more easily unmarshalled
+// into proper entity objects.
+//
+// Sample input:
+// {
+// 	"fields": [
+//         {
+//             "algorithm": {
+//                 "default": "round-robin",
+//                 "one_of": ["consistent-hashing", "least-connections", "round-robin"],
+//                 "type": "string"
+//             }
+//         }, {
+//             "hash_on": {
+//                 "default": "none",
+//                 "one_of": ["none", "consumer", "ip", "header", "cookie"],
+//                 "type": "string"
+//             }
+//         }, {
+//             "hash_fallback": {
+//                 "default": "none",
+//                 "one_of": ["none", "consumer", "ip", "header", "cookie"],
+//                 "type": "string"
+//             }
+//         },
+//   ...
+// }
+//
+// Sample output:
+// {
+// 	"algorithm": "round-robin",
+// 	"hash_on": "none",
+// 	"hash_fallback": "none",
+//  ...
+// }
+func flattenDefaultsSchema(schema gjson.Result) map[string]interface{} {
+	value := schema.Get("fields")
+	results := map[string]interface{}{}
+
+	value.ForEach(func(key, value gjson.Result) bool {
+		// get the key name
+		ms := value.Map()
+		fname := ""
+		for k := range ms {
+			fname = k
+			break
+		}
+
+		ftype := value.Get(fname + ".type")
+		if ftype.String() == "record" {
+			newSubConfig := flattenDefaultsSchema(value.Get(fname))
+			results[fname] = newSubConfig
+			return true
+		}
+		value = value.Get(fname + ".default")
+		if value.Exists() {
+			results[fname] = value.Value()
+		} else {
+			results[fname] = nil
+		}
+		return true
+	})
+
+	return results
+}
+
+func getDefaultsObj(schema map[string]interface{}) ([]byte, error) {
+	jsonSchema, err := json.Marshal(&schema)
+	if err != nil {
+		return nil, err
+	}
+	gjsonSchema := gjson.ParseBytes((jsonSchema))
+	defaults := flattenDefaultsSchema(gjsonSchema)
+	jsonSchemaWithDefaults, err := json.Marshal(&defaults)
+	if err != nil {
+		return nil, err
+	}
+	return jsonSchemaWithDefaults, nil
+}
+
+// FillUpstreamsDefaults ingests upstreams' defaults from their schema.
+func FillUpstreamsDefaults(upstream *Upstream, schema map[string]interface{}) error {
```

Comment:
Since the code doesn't actually deal with the type of the first argument, can we have a generic function that takes in an "interface{}" and does the right thing based on type of the input.
You will need to use a "type switch" in Go.

Reply:
🧙‍♂️